internal/tlssession/tlssession.go

```diff
@@ -88,8 +89,17 @@ func (ws *workersState) worker() {
 		case notif := <-ws.notifyTLS:
 			if (notif.Flags & model.NotificationReset) != 0 {
 				if err := ws.tlsAuth(); err != nil {
+
+					// TODO(ainghazal): pass the failure to the tracer too.
+
+					if errors.Is(err, ErrBadCA) {
+						ws.sessionManager.Failed <- session.NewFailure(err)
```

Comment:
Is there a reason why you need to wrap this in a special `Failure` struct?

Since go 1.13 there is also the ability to [wrap errors](https://go.dev/blog/go1.13-errors) using the `%w` formatter, which perhaps applies your use-case.

Reply:
The intention was to be able to extend it in the future for different types of failures, but you're right YAGNI for now; I'll refactor to just pass the error for now.